src/audio/up_down_mixer/up_down_mixer.c

```diff
@@ -0,0 +1,524 @@
+
+	if (cd->downmix_coefficients) {
+		ret = memcpy_s(&custom_coeffs, sizeof(custom_coeffs), downmix_coefficients,
+			       sizeof(int32_t) * UP_DOWN_MIX_COEFFS_LENGTH);
+
+		if (ret < 0)
+			return ret;
+
+		cd->downmix_coefficients = custom_coeffs;
+
+		return 0;
+	}
+
+	switch (format->ch_cfg) {
+	case IPC4_CHANNEL_CONFIG_MONO:
+	case IPC4_CHANNEL_CONFIG_STEREO:
+	case IPC4_CHANNEL_CONFIG_2_POINT_1:
+	case IPC4_CHANNEL_CONFIG_DUAL_MONO:
+		cd->downmix_coefficients = k_lo_ro_downmix32bit;
+		break;
+	case IPC4_CHANNEL_CONFIG_3_POINT_0:
+	case IPC4_CHANNEL_CONFIG_3_POINT_1:
+		if (format->depth == IPC4_DEPTH_16BIT)
+			cd->downmix_coefficients = k_half_scaled_lo_ro_downmix16bit;
+		else
+			cd->downmix_coefficients = k_half_scaled_lo_ro_downmix32bit;
+		break;
+	case IPC4_CHANNEL_CONFIG_QUATRO:
+		if (out_chanel_config == IPC4_CHANNEL_CONFIG_MONO) {
+			cd->downmix_coefficients = (format->depth == IPC4_DEPTH_16BIT) ?
+						    k_quatro_mono_scaled_lo_ro_downmix16bit :
+						    k_quatro_mono_scaled_lo_ro_downmix32bit;
+		} else { /*out_channel_config == IPC4_CHANNEL_CONFIG_STEREO*/
+			cd->downmix_coefficients = (format->depth == IPC4_DEPTH_16BIT) ?
+						    k_half_scaled_lo_ro_downmix16bit :
+						    k_half_scaled_lo_ro_downmix32bit;
+		}
+		break;
+	case IPC4_CHANNEL_CONFIG_4_POINT_0:
+		if (format->depth == IPC4_DEPTH_16BIT) {
+			cd->downmix_coefficients = k_scaled_lo_ro_downmix16bit;
+		} else {
+			if (out_chanel_config == IPC4_CHANNEL_CONFIG_5_POINT_1)
+				cd->downmix_coefficients = k_lo_ro_downmix32bit;
+			else
+				cd->downmix_coefficients = k_scaled_lo_ro_downmix32bit;
+		}
+		break;
+	case IPC4_CHANNEL_CONFIG_5_POINT_0:
+	case IPC4_CHANNEL_CONFIG_5_POINT_1:
+	case IPC4_CHANNEL_CONFIG_7_POINT_1:
+		cd->downmix_coefficients = k_scaled_lo_ro_downmix32bit;
+		break;
+	default:
+		comp_err(dev, "set_downmix_coefficients(): invalid channel config.");
+		return -EINVAL;
+		break;
+	}
+
+	return 0;
+}
+
+static up_down_mixer_routine select_mix_out_stereo(struct comp_dev *dev,
+						   const struct ipc4_audio_format *format)
+{
+	if (format->depth == IPC4_DEPTH_16BIT) {
+		switch (format->ch_cfg) {
+		case IPC4_CHANNEL_CONFIG_MONO:
+			return shiftcopy16bit_mono;
+		case IPC4_CHANNEL_CONFIG_DUAL_MONO:
+		case IPC4_CHANNEL_CONFIG_STEREO:
+			return shiftcopy16bit_stereo;
+		case IPC4_CHANNEL_CONFIG_2_POINT_1:
+		case IPC4_CHANNEL_CONFIG_3_POINT_0:
+		case IPC4_CHANNEL_CONFIG_3_POINT_1:
+		case IPC4_CHANNEL_CONFIG_QUATRO:
+		case IPC4_CHANNEL_CONFIG_4_POINT_0:
+		case IPC4_CHANNEL_CONFIG_5_POINT_0:
+			return downmix16bit;
+		case IPC4_CHANNEL_CONFIG_5_POINT_1:
+			return downmix16bit_5_1;
+		case IPC4_CHANNEL_CONFIG_INVALID:
+		default:
+			comp_err(dev, "select_mix_out_stereo(): invalid channel config.");
+			/*
+			 * This is a strange situation. We will allow to process it
+			 * in the release code (hoping for the best), but will assert
+			 * in debug to double check if everything is ok
+			 */
+			return downmix16bit;
```

Comment:
Can we return NULL here so that caller can pass the error up the stack to the host.